## Tools for demultiplexed single-cell HiC data
### Author: Maoxu Wang, Xie Lab, Peking University

### 1. Install
The following packages are required:
- python >=3.6
- python-levenshtein>=0.12.0
- pysam>=0.16.0.1 
- multiprocess>=0.70.6.1
- pandas>=0.23.4
- pybktree==1.1
- cooler>=0.9.1
```shell
    git clone https://github.com/MaoxuWang/dscHiCtools.git
    cd dscHiCtools/
    python setup.py install
```

### 2. Map barcode to whitelist & Add it to read name
R2 file (or I2 file) records the cell barcode information, add it to the read name of R1 and R3
```shell
    dscHiCtools mapBarcode \
        --read1 $Read1 \
        --index $Read2 \
        --read2 $Read3 \
        --outdir $outdir/countCB \
        --sampleName $sampleid \
        --threads $threads \
        --n_lines -1 \ # -1 means all of reads 
        --chemistry "atac" # atac or multiome
```

#### outputs:
- .R1.barcoded.fastq.gz
- .R2.barcoded.fastq.gz
- .R3.barcoded.fastq.gz

### 3. Add 'CB' tag to bam file
add "CB" tag to make it compatible for main downstream analysis 
```shell
    dscHiCtools addTag \
        --input_bam $outdir/align/${sampleid}.sorted.bam \
        --output_bam $outdir/align/${sampleid}.cellbarcoded.bam \
        --threads $SLURM_CPUS_PER_TASK \
        --samtools $samtools # default: Your Environment Variables
```
Note that you have to sort `samtools sort in.bam > out.sorted.bam` the input bam file.
#### outputs:
- output_bam with 'CB' tags added

### 4. Filter barcodes from bam file 
save barcodes listed to bam
```shell
    dscHiCtools filterBarcode \
        --input_bam $outdir/${sampleid}.cellbarcoded.bam \
        --output_bam $outdir/${sampleid}.filtered.bam \
        --barcode_file barcodes.tsv.gz \
        --threads $threads \
        --samtools $samtools # default: Your Environment Variables
```
#### outputs
- .filtered.bam


### 5. Split into single cells
split bam into different bam files by cell barcodes
```shell
    dscHiCtools splitBarcode \
        --input_bam INPUT_BAM \
        --outdir "." \
        --barcode_file cellbarcode.tsv.gz \ # cellbarcodes to split
        --threads $threads \
        --samtools $samtools # default: Your Environment Variables
```
#### outputs
- .sc.bam


### 6. Merge sub-cluster single cells to pseudo-bulk
merge and then convet to .hic and .mcool file 
```shell

    dscHiCtools sub2cool \
        --input_pairs mESC.contacts.pairs.txt.gz \
        --outdir "." \
        --metadata cellbarcode_group.metadata.txt \
        --threads $threads \
        --species mm10 \
        --resolution 1000000 50000 \ ## separated by " "
        -n \ ## if set, don't normalize matrices 
        --cooler /share/home/wangmx/anaconda3/envs/cooler/bin/cooler \ # cooler executable path
        --hic2cool /share/home/wangmx/anaconda3/envs/cooler/bin/hic2cool \ # hic2cool executable path
        --juicer "java -Xmx64G -jar /share/home/wangmx/software/juicer/scripts/common/juicer_tools_1.22.01.jar" \ # juicer executable path
```
#### Metadata format ("\t" separated)
```shell
cellbarcode     group
AACTGGTAGAATATCG        HCT-116
TTTGGTTAGTTACCAC        HCT-116
ATCCCTGGTTAACCGT        HCT-116
AGGCGTCAGAAATCTG        eHAP
ATCCTCGAGTACAGTA        HCT-116
ATTGTCTTCAACTTGG        HCT-116
AAATGAGAGGTCTTTG        HCT-116
```
#### outputs
- sub_cluster.contacts.pairs.txt.gz (4DN format)
- sub_cluster.hic
- sub_cluster.mcool

### 7. scA/B value calculation
calculate scA/B matrix for dscHiC data
```shell
    dscHiCtools scAB \
        --input_contact_pairs $outdir/$sampleid.contacts.paris.txt.gz \
        --output $outdir/$sampleid.AB_value.txt.gz \
        --weighted True \
        --ref_CpG $ref_cpg \
        --barcode_kept $outdir/cellbarcodes.txt \
        --input_format "scVI-3d" \
        --cores $SLURM_CPUS_PER_TASK 
```
#### input_format
- scVI-3d, use `.scVI_imputed.contacts.paris.txt.gz`
- raw (as Tan 2018 described) use `.contacts.paris.txt.gz`
- higashi_imputed, use `.hdf5` generated by higashi

#### ref_cpg file format
```shell
chr1    0       0.016217
chr1    1000000 0.037035
chr1    2000000 0.030277
chr1    3000000 0.028140
chr1    4000000 0.013604
chr1    5000000 0.013502
chr1    6000000 0.022897
chr1    7000000 0.015870
chr1    8000000 0.015044
chr1    9000000 0.020502
chr1    10000000        0.018762
```
- This file can be generated by dip-C (https://github.com/tanlongzhi/dip-c)
#### outputs
- sub_cluster.contacts.pairs.txt.gz (4DN format)
- sub_cluster.hic
- sub_cluster.mcool


### 8. Split contacts pairs file to single-cell level
Split raw contacts pairs file to small single-cell level files
```shell
    dscHiCtools splitContacts \
        --input_contacts $outdir/$sampleid.contacts.paris.txt.gz \
        --outdir $outdir/sc_contacts \
        --threads $threads \
        --barcode_file $outdir/cellbarcodes.txt
```
barcode_file is a txt file stores single-cell barcodes of interest.